### ISSUE TITLE
Wrong handling of `exclusiveMinimum` and `exclusiveMaximum`

### ISSUE THREAD

**Issue by Bomret** [atlantishealthcare/swagger-model-validator#84]
The swagger spec states in section 5.1.3, that if `exclusiveMinimum` is present, the value must be greater than the minimum. For `exclusiveMaximum` the value must be smaller than the maximum. But this library just adds or subtracts `1` if it is present which is against the spec.

**Example:**

- minimum: 0
- exclusiveMinimum: true

According to the spec, `0.00000000001` would be ok, but this library says it has to be at least `1.00000000001`, which is wrong.

I haven't checked other use cases of `exclusiveMin/Max` (like dates).

If an `exclusiveMinimum` is there, the value should just be checked via `value > minimum` and `value >= minimum` if not. Vice versa for `exclusiveMaximum`.

**Comment by brodyd** [comment]
Could you please link to that reference for the spec.  I can't find any reference to section numbers so I can't check 5.1.3.

My reading of the swagger specification shows that the keywords exclusiveMinimum and exclusiveMaximum are as per the JSON Schema and that shows them as numbers indicating an exclusive range and not as a Boolean that changes behaviour based on the other minimum and maximum fields.

However, while I was verifying things worked as expected I discovered a nasty bug when exclusiveMinimum or exclusiveMaximum are zero and the type is a float or a double.  So I have refactored to check using the base values instead as you suggest.  

I'll have a patch out shortly.

**Comment by Bomret** [comment]
Here you are: https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.1.3

`minimum`, `maximum` are numbers and their `exclusive` counterparts are booleans that change the behavior. This is also implemented in this lib but there is a hardcoded `minimum + 1` which is not correct. If `minimum` is 0 and `exclusiveMinimum` is true then a value of 0 is invalid, but any value greater than 0 is valid, which includes e.g. 0.000000000001

**Comment by brodyd** [comment]
I think that is the old draft.

The current draft is http://json-schema.org/latest/json-schema-validation.html#rfc.section.6.3

Which says 

> The value of "exclusiveMaximum" MUST be number, representing an exclusive upper limit for a numeric instance.
> 
> If the instance is a number, then the instance is valid only if it has a value strictly less than (not equal to) "exclusiveMaximum".

Which is what has been implemented.  Although as mentioned above there was a bug when the value was zero which has now be fixed - pretty much by moving to check as you suggested.

**Comment by Bomret** [comment]
Yeah, but the current draft is not applicable for OpenAPI (swagger) spec 2.0 which this library is about. Spec 2.0 (https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md) explicitely links the explanations for e.g. `exclusiveMinimum` to section 5.3.1 of the draft version I linked above. So in swagger 2.0 `exclusiveMinimum` and `exclusiveMaximum` ought to be booleans, not numbers. This was a breaking change between OpenAPI 2.0 and 3.0.

**Comment by brodyd** [comment]
You are bang on correct.  That is a problem...  I'm not sure how we are going to be able to handle that.

**Comment by Bomret** [comment]
Well this lib explicitly states that it supports OpenAPI spec 2.0. I can take a look at the current validation code next week after my vacation. But maybe it would be best to release a new version as a major version indicating a breaking change between the previous releases and the new one.

**Comment by brodyd** [comment]
We'll see if we can't support both...

**Comment by Bomret** [comment]
It would be nice if this lib would be able to support OpeanAPI 3.0 too, but this is an entirely different beast, I think.

**Comment by brodyd** [comment]
I've pushed out a 2.2 release with these changes.

I suspect this was a change from behaviour in Swagger 1.1 and no body has noticed it in swagger 2.0.  It appears to have been changed back for swagger 3.0 (or at least the underlying JSON schema has)

**Comment by Bomret** [comment]
Cool! Need to check it out next week.

**Comment by brodyd** [comment]
I'll leave this open until you have had a chance to check it out.  Thanks for your input.

**Comment by Bomret** [comment]
It seems as if the newest version contains two type errors that lead to failing validation: Lines `711` and `715` (function `validateMinMaxValue`) of the `modelValidator` module contain a `maxValue.ToString()`, but in JavaScript the function is called `toString` with a lowercase `t`. This leads to a `TypeError` being thrown.

**Comment by Bomret** [comment]
It seems that a large part of the `validateMinMaxValue` function is broken now.

Spec:
```yaml
Foo:
  properties:
    value:
      type: number
      minimum: -90
      maximum: 90
```

Test object:
```
{
  value: 53.252525
}
```

is marked as invalid because the `if-else` block from line `760` to `768` only contains error branches. The `return null` at the end in line `770` is never hit.

**Comment by brodyd** [comment]
I've fixed this in 2.2.1
